src/modules/0.0.14/aws_lambda/index.ts

```diff
@@ -0,0 +1,204 @@
+import { AwsIamModule } from '../aws_iam'
+
+const base64ToUint8Array = (base64: string) => {
+  const buf = Buffer.from(base64, 'base64');
+  const bytes = new Uint8Array(buf);
+  return bytes;
+}
+
+const lambdaFunctionMapper = async (fn: GetFunctionResponse, ctx: Context) => {
+  const out = new LambdaFunction();
+  out.architecture = fn.Configuration?.Architectures?.pop() as Architecture;
+  out.description = fn.Configuration?.Description;
+  out.environment = fn.Configuration?.Environment?.Variables;
+  out.handler = fn.Configuration?.Handler;
+  out.memorySize = fn.Configuration?.MemorySize;
+  out.name = fn.Configuration?.FunctionName ?? 'not possible?';
+  out.packageType = fn.Configuration?.PackageType as PackageType;
+  try {
+    const roleName = AwsIamModule.utils.roleNameFromArn(fn.Configuration?.Role);
+    out.role = await AwsIamModule.mappers.role.db.read(ctx, roleName) ??
+      await AwsIamModule.mappers.role.cloud.read(ctx, roleName);
+    if (!out.role) return undefined;
+  } catch (_e: any) {
+    // TODO: look for role not found error Code
+    return undefined;
+  }
+  out.runtime = fn.Configuration?.Runtime as Runtime;
+  out.tags = fn.Tags;
+  out.version = fn.Configuration?.Version;
+  out.arn = fn.Configuration?.FunctionArn;
+  return out;
+}
+
+const updateableFunctionFieldsEq = (a: LambdaFunction, b: LambdaFunction) => {
+  return Object.is(a.role.roleName, b.role.roleName) &&
+    Object.is(a.handler, b.handler) &&
+    Object.is(a.memorySize, b.memorySize) &&
+    Object.is(a.description, b.description) &&
+    isEqual(a.environment, b.environment) &&
+    Object.is(a.runtime, b.runtime);
+}
+
+const updateableTagsEq = (a: LambdaFunction, b: LambdaFunction) => {
+  return isEqual(a.tags, b.tags);
+}
+
+const restorableFunctionFieldsEq = (a: LambdaFunction, b: LambdaFunction) => {
+  return Object.is(a.arn, b.arn) &&
+    Object.is(a.packageType, b.packageType) &&
+    Object.is(a.architecture, b.architecture) &&
+    Object.is(a.version, b.version);
+}
+
+const updateableCodeFieldsEq = (a: LambdaFunction, b: LambdaFunction) => {
+  return Object.is(a.zipB64, b.zipB64);
+}
+
+export const AwsLambdaModule: Module2 = new Module2({
+  ...metadata,
+  mappers: {
+    lambdaFunction: new Mapper2<LambdaFunction>({
+      entity: LambdaFunction,
+      equals: (a: LambdaFunction, b: LambdaFunction) => updateableFunctionFieldsEq(a, b) &&
+        updateableTagsEq(a, b) &&
+        restorableFunctionFieldsEq(a, b) &&
+        updateableCodeFieldsEq(a, b),
+      source: 'db',
+      cloud: new Crud2({
+        create: async (es: LambdaFunction[], ctx: Context) => {
+          const client = await ctx.getAwsClient() as AWS;
+          const out = [];
+          for (const e of es) {
+            const input: CreateFunctionCommandInput = {
+              FunctionName: e.name,
+              Description: e.description,
+              Code: {
+                ZipFile: base64ToUint8Array(e.zipB64 ?? ''),
+              },
```

Comment:
We should probably skip and/or drop the creation of a lambda function without the base64-encoded zip file defined.

Also, add a TODO here to validate that the b64 contents are actually a zip file?

Reply:
The options I see here on create are either `throw` if no zip, or `db.delete` the row. I'd go with the `throw` option at least for now to let users know what's wrong.

Reply:
Regarding the TODO suggestion 👍🏼